### ISSUE TITLE
fuzzing: compilation error ‘__builtin_mul_overflow’ does not have pointer to integer type

### ISSUE THREAD

**Issue by naumenkogs** [bitcoin/bitcoin#18389]
I'm using Debian, compiling current master (5bf45fe2a9642f8ae8f8a12bcbf8f8b4770421ad) after configure with `./configure --disable-shared --enable-tests --enable-fuzz CC=${AFLPATH}/afl-clang CXX=${AFLPATH}/afl-clang++`.

Getting this error:
```[+] Instrumented 442 locations (64-bit, hardened mode, ratio 100%).
  CXX      test/fuzz/test_fuzz_key-fuzz.o
afl-cc 2.52b by <lcamtuf@google.com>
test/fuzz/integer.cpp:152:27: error: default initialization of an object of const type 'const SigningProvider' without a
      user-provided default constructor
    const SigningProvider store;
                          ^
                               {}
1 error generated.
Makefile:14931: recipe for target 'test/fuzz/test_fuzz_integer-integer.o' failed
```

If I configure for gcc, I get this error:
```
test/fuzz/multiplication_overflow.cpp: In instantiation of ‘void {anonymous}::TestMultiplicationOverflow(FuzzedDataProvider&) [with T = bool]’:
test/fuzz/multiplication_overflow.cpp:41:58:   required from here
test/fuzz/multiplication_overflow.cpp:21:75: error: argument 3 in call to function ‘__builtin_mul_overflow’ does not have pointer to integer type
     const bool is_multiplication_overflow_builtin = __builtin_mul_overflow(i, j, &result_builtin);
                                                                           ^
Makefile:15155: recipe for target 'test/fuzz/test_fuzz_multiplication_overflow-multiplication_overflow.o' failed
```

Funny enough, it works on my macbook locally :)
cc @practicalswift

**Comment by MarcoFalke** [comment]
cc @practicalswift https://github.com/bitcoin/bitcoin/pull/18353#discussion_r393326105

**Comment by practicalswift** [comment]
Thanks for the ping! I'll fix :)

**Comment by practicalswift** [comment]
@naumenkogs Thanks for reporting. The `__builtin_mul_overflow` issue is fixed by #18393. 

I cannot reproduce the "default initialization of an object of const type …" issue. What versions of AFL and gcc are you using? :)

**Comment by MarcoFalke** [comment]
> I cannot reproduce the "default initialization of an object of const type …" issue. What versions of AFL and gcc are you using? :)

If I had to guess, it is probably an older version of gcc, like 4.8 or 5.3

**Comment by naumenkogs** [comment]
>I cannot reproduce the "default initialization of an object of const type …" issue. What versions of AFL and gcc are you using? :)

This issue happens with clang:
clang version 3.8.0-2ubuntu4 (tags/RELEASE_380/final)
AFL I'm using is 2.52b

**Comment by practicalswift** [comment]
@naumenkogs Are you able to reproduce it under a newer version of Clang? :)

I'm usually running Clang 10 compiled from trunk but I now tested against Clang 4 which is the oldest version I have around. I'm unable to reproduce under Clang 4.

**Comment by MarcoFalke** [comment]
@practicalswift I think @naumenkogs is using debian stretch: https://packages.debian.org/stretch/clang

**Comment by naumenkogs** [comment]
After updating clang, the clang issue got fixed. Should this be mentioned in docs? I think there's no such requirement for Bitcoin Core. 

As for the gcc issue — this still applies #18393 even after compilers update, waiting for the solution in that PR :)

**Comment by naumenkogs** [comment]
Fixed in 7c1ac70
